src/protobuf.zig

```diff
@@ -723,81 +687,96 @@ fn decode_packed_list(slice: []const u8, comptime list_type: ListType, comptime
-                            .RawValue => |value| {
-                                try @field(result, field.name).append(decode_fixed_raw(child_type, value));
-                            },
-                            .Slice => |slice| {
-                                try decode_packed_list(slice, list_type, child_type, &@field(result, field.name));
-                            },
-                        }
-                    },
-                    .SubMessage, .String => switch (extracted_data.data) {
-                        .Slice => |slice| try decode_list(slice, list_type, child_type, &@field(result, field.name), allocator),
-                        .RawValue => @panic("TODO: TEST Invalid data"),
-                    },
-                }
+            switch (list_type) {
+                .Varint => |varint_type| {
+                    switch (extracted_data.data) {
+                        .RawValue => |value| try @field(result, field.name).append(decode_varint_value(child_type, varint_type, value)),
+                        .Slice => |slice| try decode_packed_list(slice, list_type, child_type, &@field(result, field.name)),
+                    }
+                },
+                .FixedInt => |_| {
+                    switch (extracted_data.data) {
+                        .RawValue => |value| try @field(result, field.name).append(decode_fixed_value(child_type, value)),
+                        .Slice => |slice| try decode_packed_list(slice, list_type, child_type, &@field(result, field.name)),
+                    }
+                },
+                .SubMessage => switch (extracted_data.data) {
+                    .Slice => |slice| try @field(result, field.name).append(try child_type.decode(slice, allocator)),
+                    .RawValue => return error.InvalidInput,
+                },
+                .String => switch (extracted_data.data) {
+                    .Slice => |slice| try @field(result, field.name).append(slice),
+                    .RawValue => return error.InvalidInput,
+                },
             }
         },
-        .OneOf => |_| {
-            @compileError("Can not decode OneOf fields yet");
+        .OneOf => |one_of| {
+            // the following code:
+            // 1. creates a compile time for iterating over all `one_of._union_desc` fields
+            // 2. when a match is found, it creates the union value in the `field.name` property of the struct `result`. breaks the for at that point
+            const desc_union = one_of._union_desc;
+            inline for (@typeInfo(one_of).Union.fields) |union_field| {
+                const v = @field(desc_union, union_field.name);
+                if (is_tag_known(v, extracted_data)) {
+                    var value = try decode_value(union_field.type, v.ftype, extracted_data, allocator);
+                    @field(result, field.name) = @unionInit(one_of, union_field.name, value);
+                }
+            }
```

Comment:
Thank you for finding a solution for this!  I remember trying decoding oneOfs at some point but i'm not sure 
`@unionInit` existed yet. Very happy about this.